File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/passes/ReplaceDeclarationsHavingInitializersWithAssignments.java
package ro.pub.cs.diploma.passes;

import com.intellij.psi.*;
import com.intellij.refactoring.util.RefactoringUtil;
import org.jetbrains.annotations.NotNull;
import ro.pub.cs.diploma.NameManager;
import ro.pub.cs.diploma.RecursionUtil;
import ro.pub.cs.diploma.Util;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class ReplaceDeclarationsHavingInitializersWithAssignments implements Pass<PsiCodeBlock, List<PsiDeclarationStatement>, Object> {
  @NotNull private final PsiMethod myMethod;
  @NotNull private final NameManager myNameManager;
  @NotNull private final PsiElementFactory myFactory;

  private ReplaceDeclarationsHavingInitializersWithAssignments(@NotNull final PsiMethod method,
                                                               @NotNull final NameManager nameManager,
                                                               @NotNull final PsiElementFactory factory) {
    myMethod = method;
    myNameManager = nameManager;
    myFactory = factory;
  }

  @NotNull
  public static ReplaceDeclarationsHavingInitializersWithAssignments getInstance(@NotNull final PsiMethod method,
                                                                                 @NotNull final NameManager nameManager,
                                                                                 @NotNull final PsiElementFactory factory) {
    return new ReplaceDeclarationsHavingInitializersWithAssignments(method, nameManager, factory);
  }

  @Override
  public List<PsiDeclarationStatement> collect(PsiCodeBlock block) {
    final List<PsiDeclarationStatement> declarations = new ArrayList<>();
    block.accept(new JavaRecursiveElementWalkingVisitor() {
      @Override
      public void visitDeclarationStatement(PsiDeclarationStatement statement) {
        if (RecursionUtil.hasToBeSavedOnStack(statement, myMethod)) {
          declarations.add(statement);
        }
      }

      @Override
      public void visitClass(PsiClass aClass) {
      }

      @Override
      public void visitLambdaExpression(PsiLambdaExpression expression) {
      }
    });
    return declarations;
  }

  private static Stream<String> getVariablesStream(@NotNull final PsiDeclarationStatement statement, @NotNull final String frameVarName) {
    return Arrays
      .stream(statement.getDeclaredElements())
      .filter(element -> element instanceof PsiLocalVariable)
      .map(element -> (PsiLocalVariable)element)
      .filter(PsiVariable::hasInitializer)
      .map(variable -> frameVarName + "." + variable.getName() + " = " +
                       RefactoringUtil.convertInitializerToNormalExpression(variable.getInitializer(), variable.getType()).getText());
  }

  @Override
  public Object transform(List<PsiDeclarationStatement> declarations) {
    for (final PsiDeclarationStatement statement : declarations) {
      final PsiElement parent = statement.getParent();
      final Stream<String> stream = getVariablesStream(statement, myNameManager.getFrameVarName());
      if (parent instanceof PsiForStatement) {
        statement.replace(Util.statement(myFactory, stream.collect(Collectors.joining(",")) + ";"));
        continue;
      }
      final PsiCodeBlock parentBlock = (PsiCodeBlock)parent;
      PsiElement anchor = statement;
      for (final String string : stream.collect(Collectors.toList())) {
        anchor = parentBlock.addAfter(Util.statement(myFactory, string + ";"), anchor);
      }
      statement.delete();
    }
    return null;
  }
}

File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/passes/ReplaceDeclarationsHavingInitializersWithAssignments.kt
package ro.pub.cs.diploma.passes

import com.intellij.psi.*
import com.intellij.refactoring.util.RefactoringUtil
import ro.pub.cs.diploma.NameManager
import ro.pub.cs.diploma.containsInScopeRecursiveCallsTo
import ro.pub.cs.diploma.statement
import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream

class ReplaceDeclarationsHavingInitializersWithAssignments(private val myMethod: PsiMethod,
                                                           private val myNameManager: NameManager,
                                                           private val myFactory: PsiElementFactory) : Pass<PsiCodeBlock, List<PsiDeclarationStatement>, Any> {

  override fun collect(block: PsiCodeBlock): List<PsiDeclarationStatement> {
    val declarations = ArrayList<PsiDeclarationStatement>()
    block.accept(object : JavaRecursiveElementWalkingVisitor() {
      override fun visitDeclarationStatement(statement: PsiDeclarationStatement) {
        if (statement.containsInScopeRecursiveCallsTo(myMethod)) {
          declarations.add(statement)
        }
      }

      override fun visitClass(aClass: PsiClass) {}

      override fun visitLambdaExpression(expression: PsiLambdaExpression) {}
    })
    return declarations
  }

  override fun transform(declarations: List<PsiDeclarationStatement>): Any? {
    for (statement in declarations) {
      val parent = statement.parent
      val stream = getVariablesStream(statement, myNameManager.frameVarName)
      if (parent is PsiForStatement) {
        statement.replace(myFactory.statement(stream.collect(Collectors.joining(",")) + ";"))
        continue
      }
      val parentBlock = parent as PsiCodeBlock
      var anchor: PsiElement = statement
      for (string in stream.collect(Collectors.toList())) {
        anchor = parentBlock.addAfter(myFactory.statement( "$string;"), anchor)
      }
      statement.delete()
    }
    return null
  }

  private fun getVariablesStream(statement: PsiDeclarationStatement, frameVarName: String): Stream<String> {
    return Arrays
        .stream(statement.declaredElements)
        .filter { element -> element is PsiLocalVariable }
        .map { element -> element as PsiLocalVariable }
        .filter({ it.hasInitializer() })
        .map { variable ->
          "$frameVarName.${variable.name} = ${RefactoringUtil.convertInitializerToNormalExpression(variable.initializer, variable.type).text}"
        }
  }
}
